src/lib/homestore_backend/snapshot_receive_handler.cpp

```diff
@@ -0,0 +1,213 @@
+    auto status = homestore::data_service().alloc_blks(
+        sisl::round_up(sizeof(shard_sb), homestore::data_service().get_blk_size()), hints, blk_id);
+    if (status != homestore::BlkAllocStatus::SUCCESS) {
+        LOGE("Failed to allocate blocks for shard {}", shard_meta.shard_id());
+        return ALLOC_BLK_ERR;
+    }
+
+    const auto ret = homestore::data_service()
+                         .async_write(r_cast< char const* >(&shard_sb), sizeof(shard_sb), blk_id)
+                         .thenValue([&blk_id](auto&& err) -> BlobManager::AsyncResult< blob_id_t > {
+                             // TODO: do we need to update repl_dev metrics?
+                             if (err) {
+                                 LOGE("Failed to write shard info to blk_id: {}", blk_id.to_string());
+                                 return folly::makeUnexpected(BlobError(BlobErrorCode::REPLICATION_ERROR));
+                             }
+                             LOGD("Shard info written to blk_id:{}", blk_id.to_string());
+                             return 0;
+                         })
+                         .get();
+    if (ret) {
+        LOGE("Failed to write shard info of shard_id {} to blk_id:{}", shard_meta.shard_id(), blk_id.to_string());
+        return WRITE_DATA_ERR;
+    }
+
+    // Now let's create local shard
+    home_obj_.local_create_shard(shard_sb.info, chunk_id, blk_id.blk_count());
+    shard_cursor_ = shard_meta.shard_id();
+    blob_cursor_ = 0;
+    cur_batch_num_ = 0;
+    return 0;
+}
+
+int HSHomeObject::SnapshotReceiveHandler::process_blobs_snapshot_data(ResyncBlobDataBatch const& data_blobs,
+                                                                      const snp_batch_id_t batch_num) {
+    cur_batch_num_ = batch_num;
+    for (unsigned int i = 0; i < data_blobs.blob_list()->size(); i++) {
+        const auto blob = data_blobs.blob_list()->Get(i);
+
+        // Skip deleted blobs
+        if (blob->state() == static_cast< uint8_t >(ResyncBlobState::DELETED)) {
+            LOGD("Skip deleted blob_id:{}", blob->blob_id());
+            continue;
+        }
+
+        // Check duplication to avoid reprocessing
+        shared< BlobIndexTable > index_table;
+        {
+            std::shared_lock lock_guard(home_obj_._pg_lock);
+            auto iter = home_obj_._pg_map.find(pg_id_);
+            RELEASE_ASSERT(iter != home_obj_._pg_map.end(), "PG not found");
+            index_table = dynamic_cast< HS_PG* >(iter->second.get())->index_table_;
+        }
+        RELEASE_ASSERT(index_table != nullptr, "Index table instance null");
+        if (home_obj_.get_blob_from_index_table(index_table, shard_cursor_, blob->blob_id())) {
+            LOGD("Skip already persisted blob_id:{}", blob->blob_id());
+            continue;
+        }
+
+        auto blob_data = blob->data()->Data();
+        auto header = r_cast< BlobHeader const* >(blob_data);
+        if (header->valid()) {
```

Comment:
nit: `if (!header->valid())`. 
And we consider the blob with invalid header as corrupted as well, so should not return error in this branch.

Reply:
Nice catch, moved the header validation into the if statement body.